### PR TITLE
fix(NotificationDrawer) - Dropdown not overlapping Clear/Read Section

### PR DESCRIPTION
**Pull request by gilad215**
**What**:
Fixing Dropdown not overlapping Clear/Read Section.
More Info: patternfly/patternfly#1039

After wrestling with the CSS finally found a solution for this issue while keeping flex for the outer part of the drawer 🎉 

**Link to Storybook**:
[Storybook](https://rawgit.com/gilad215/patternfly-react/storybook/fix/notificationdrawer/index.html?knob-is%20Drawer%20Expandable=true&selectedKind=NotificationDrawer&selectedStory=Stateful%20Drawer%20Wrapper&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

### CODE REVIEW

**Comment by LHinson** [comment]
nice, thanks @gilad215 for the contribution. 

@jeff-phillips-18 since this a CSS update, should this be sent to PF-Core first and then updated in React?

**Comment by jeff-phillips-18** [comment]
@LHinson This fix won't be appropriate for https://github.com/patternfly/patternfly as this version of the Notification Drawer uses flexbox whereas the core version is still using fixed height computations.

If this is desireable in PF core, then we should add an issue to https://github.com/patternfly/patternfly to include a flexbox version of the notification drawer. Not a fix to the current as it would definitely be breaking and would require the application to handle open/close rather than using bootstrap's accordion functionality.

File: less/notificationdrawer.less [hunk 1]
```diff
@@ -23,19 +23,26 @@
...
   .drawer-pf-action {
-    flex: none;
+    position: sticky;
+    position: -webkit-sticky;
+    position: -moz-sticky;
+    position: -ms-sticky;
+    position: -o-sticky;
+    bottom: 0px;
+    background-color: #ffffff;
+    border-top: 1px solid #d1d1d1;
```
Comment (jeff-phillips-18) [inline less/notificationdrawer.less]: These should use patternfly color variables.

File: less/notificationdrawer.less [hunk 2]
```diff
@@ -23,19 +23,26 @@
...
     }
   }
 
   .drawer-pf-action {
-    flex: none;
+    position: sticky;
+    position: -webkit-sticky;
+    position: -moz-sticky;
+    position: -ms-sticky;
+    position: -o-sticky;
```
Comment (jgiardino) [inline less/notificationdrawer.less]: Unfortunately, `position: sticky` is not supported in IE 11. 

@jeff-phillips-18 Do we need to include all of the vendor prefixes in this repo? Or do we use [autoprefixer](http://codeguide.patternfly.org/#css-prefixed-properties) as we do in patternfly core?

**Comment by ohadlevy** [comment]
On Tue, May 1, 2018, 9:24 PM Jenn Giardino <notifications@github.com> wrote:

> @gilad215 <https://github.com/gilad215> @jeff-phillips-18
> <https://github.com/jeff-phillips-18>
>
> This is known behavior with the notification drawer. I had opened an issue
> earlier in patternfly core: patternfly/patternfly#553
> <https://github.com/patternfly/patternfly/issues/553>.
>
> Ultimately the decision was to not fix the issue for PF 3 since you could
> still scroll to see the menu items.
>
> My preference would be to keep the css as-is until a fix is available for
> patternfly next, since:
>
>    - the css using position: sticky isn't supported in IE 11
>    - this fix only addresses the menu being cut off by the action footer,
>    but the menu is still cut off by the other panels
>    - the current implementation matches what we have in core
>
> What do you think?
>
> —
>

Why only fix it for next? There are many projects who use the current pf?
Are we already stopping support for pf3…?

**Comment by jgiardino** [comment]
Hi @ohadlevy 

> Why only fix it for next? There are many projects who use the current pf?
Are we already stopping support for pf3…?

I originally opened https://github.com/patternfly/patternfly/issues/553 almost a year and a half ago, and even then it was a known issue. So it's not a matter of no longer supporting pf3.

My guess at the reasons why it wasn't addressed then could be the level of complexity to fix and it not being an issue that blocks the user—the user can scroll to access the menu items that aren't visible.

**Comment by gilad215** [comment]
@jgiardino Hi

`position:sticky` is a problem, in some cases `position:fixed` can help with IE 11 issues but not in this case.
I do think its the correct design, the menu items should not overlap the panels (my opinion) and the drawer is already flex and not matching core so it wouldn't hurt to fix another issue.

I'm not really sure how we continue from here, PF-next creating a whole new drawer?

**Comment by gilad215** [comment]
@jeff-phillips-18 @jgiardino 
Seems like IE 11 doesn't like the Flex drawer, this is a screenshot from the master branch without any changes. 

screen shot 2018-05-06 at 12 41 53

Edge works as expected. I think having to scroll for the Read/Clear section is not an issue compared to this, Thoughts?

**Comment by jgiardino** [comment]
Oh, good catch on the IE11 issue. On the [can i use page for flexbox](https://caniuse.com/#search=flexbox), the Known Issues tab mentions the following:

> In IE10 and IE11, containers with display: flex and flex-direction: column will not properly calculate their flexed childrens' sizes if the container has min-height but no explicit height property.

And that's what we have for the expanded panel:

```
.drawer-pf.drawer-alt-pf .panel-group .panel.panel-default.expanded {
    min-height: 175px;
    flex: 1 1 auto;
    display: flex;
    flex-direction: column;
}
```

I also wonder if setting `flex: none;` on `.drawer-pf-action` is causing any issues, too.

Regarding `position: sticky`, I don't disagree with any of your points. But when I inspect your example and make the menu longer, this solution doesn't 100% solve the problem, as shown in the examples below. 

image
image

I don't really have a strong preference when comparing the two options, and am interested in what others think about this issue. 

@jeff-phillips-18 @patternfly/patternfly-react-ux – any thoughts?

File: less/notificationdrawer.less (hunk 2 above)
Comment (jeff-phillips-18) [reply less/notificationdrawer.less]: I don't believe we currently use autoprefixer.

**Comment by jeff-phillips-18** [comment]
This only prevents scrolling as much as the height of the Clear/Mark Read area. If that isn't present, it doesn't prevent any scrolling. I would prefer to leave it as is.

**Comment by jeff-phillips-18** [comment]
Sorry, I'm not sure what you mean. From what I can tell, this fix only prevents scrolling as much as the height of the mark/clear area and it causes the kebab to draw in that area when one does scroll. So, it's not fixing the scroll issue entirely but it is adding another issue.

**Comment by terezanovotna** [comment]
Looking at the storybook example, it looks great that the dropdown sits above the Clear/Read section. 

@jgiardino I think it's better to have the dropdown above it. If the user wants to Clear section, he should close the kebab and then Clear section. 

The issue I still see is that when I scroll down, the kebab icon travels all the way down, and I don't think it should be that way. 

kebab

**Comment by gilad215** [comment]
@jgiardino @jeff-phillips-18  @terezanovotna Hi, the kebab is shown only if highlighted, but the dropdown menu is shown always.
I researched other drawers, on facebook's drawer for example they disable scrolling if a dropdown menu is shown, which actually sounds reasonable. This can solve the problem, but then we need to enable scrolling inside the menu items of the dropdown in case it has a lot of actions.

**Comment by jeff-phillips-18** [comment]
For what it's worth, Github Project Board just adds height and scroll the area as we do today

http://g.recordit.co/mnQxyxS4AC.gif

**Comment by jgiardino** [comment]
@terezanovotna Please see the examples I posted above in [my previous comment](issuecomment-387479432). 

The example shown in the current storybook doesn't include enough menu items in the dropdown to produce the UI issues that I noted above. I created the examples above by modifying the runtime code in the inspector and adding options.

That said, I'm still on the fence about fixing it. I can see how in the cases where there aren't a lot of menu items, this update is better. But in the case where there are more menu items, I'd say that this is not better than what we have today. 

@jeff-phillips-18 @serenamarie125 @Rohoover Do you guys want to weigh in on this one? Would it help to compare if the storybook example had more menu items so that you can see the issue that's present with the last notification?

**Comment by jeff-phillips-18** [comment]
I'm on the fence but lean more towards leaving it how it is.

**Comment by jgiardino** [comment]
Hi @gilad215 and @ohadlevy 

I followed up with the larger patternfly team for guidance on this one. Since this solution doesn't completely solve the issue and is inconsistent with the patternfly behavior, the preference is to keep the react component as it is. But @ohadlevy previously made a suggestion that we document the fix that's included in this PR so that individual teams can choose to use this solution if they prefer. Do you think it's sufficient to include a link to this PR in the storybook example as a way of sharing this alternate solution??

**Comment by dabeng** [comment]
@gilad215 FYI, I find that there is a unnecessary scrollbar blinking on users toggling notification tabs.

qakcqhq0kb

**Comment by jeff-phillips-18** [comment]
Closing. If we want to add example CSS to storybook then we can add an issue and follow-up PR.

**Pull request closed without merging**